proposals/0111-linear-types.rst

```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
 annotation. The process for this is not yet defined (see `Unresolved
 questions`_ below for a more precise description of this issue).
 
+Let and where bindings
+----------------------
+
+*This section is written with let bindings, but all of the same applies
+to where bindings.*
+
+Note on terminology: following the Haskell 2010 report, a function
+binding is a binding of the form ``f arg1…argn =`` with at least one
+arguments. The other binding is called pattern bindings. In particular
+``let x =`` is a pattern binding (as opposed to how it is in GHC's
+implementation, where ``let x =`` would be a ``FunBind``). This
+proposal will be using the terminology “non-variable pattern”
+for patterns which aren't a single variable.
+
+Let bindings can optionally be annotated with a multiplicity
+(including a multiplicity variable):
+
+::
+
+   let
+     %p1 pat1 = rhs1
+     …
+     %pn patn = rhsn
+   in body
+
+if
+
+* The binding is not top-level
+* The binding is non-recursive
+* The binding is a pattern binding (including a simple variable)
+  ``p=e``
+* Either ``p`` is of the form ``!p'`` or ``p`` is a variable. In
+  particular neither ``x@y`` nor ``(x)`` are covered by “is a
+  variable”
+
+If there is a multiplicity annotation, the binding is not
+generalised. So ``let %m x = e in b`` and ``(\(%m x) ->  b) e`` have
+the same typing rule.
+
+When the multiplicity annotation isn't specified, the multiplicity is
+inferred, so that both
+
+::
+
+   \(%1 z) -> let !(x, y) = z in (y, x)
+   let !(x, y) = u in (x, x)
+
+are well typed. Recursive bindings, toplevel bindings and non-variable
+lazy-pattern bindings are always inferred to have multiplicity
+``Many``.
+
+Note that, in particular, function bindings are not allowed:
+
+::
+
+   -- Doesn't work
+   let %p f x y = rhs in body
+
+instead write
+
+::
+
+   -- pat can be a variable
+   let %p f = \x y -> rhs in body
+
+When typechecking ``let %p pat = rhs in body`` the ``rhs`` is consumed
+with multiplicity ``p`` and the variables of ``pat`` must be consumed
+in ``body`` with multiplicity ``p`` (if ``pat`` has some non-linear
+fields, then the variables are scaled appropriately as per
+`Constructors & pattern-matching`_). The pattern ``pat`` can be an
+arbitrary, nested, pattern.
+
+Here are a few examples that illustrate the typing rules
+
+::
+
+   -- good
+   let %1 x = u in x
+
+   -- bad
+   let %1 x = u in (x, x)
+
+   -- bad
+   let %1 x = u in let %Many y = x in …
+
+   -- good
+   let %1 (x, y) = u in (y, x)
+
+   -- bad
+   let %1 (x, y) = u in x
+
+   -- good
+   let %1 (Ur x) = u in (x, x, x)
+
+Non-variable let-bound patterns must be strict
```

Comment:
This is covered by the section _Pattern-matching_ of the (unamended part of the) proposal. `~` patterns are unrestricted (maybe we should say “unrestricting” for a pattern as it's kind of a co-expression :slightly_smiling_face: ). View patterns are unrestricted as well, but not by design, rather by ignorance of what we could do better.

In this section, I only cover what's different for lets since patterns have a slightly different semantic in let-bindings from that in case-expressions.